### ISSUE TITLE
guarding against timing based attacks

### ISSUE THREAD

**Issue by davidmarkclements** [mcollina/bloomrun#30]
if we do string equality checks, any tokens/hashes etc. used in patterns makes the system vulnerable to timing attacks

we need a constant-time equality check like so:

```js
var mismatch = 0
var len = a.length
var i
for (i = 0; i < len; ++i) {
  mismatch |= (a.charCodeAt(i) ^ b.charCodeAt(i))
}
return mismatch
```

CF: https://snyk.io/blog/node-js-timing-attack-ccc-ctf/

**Comment by davidmarkclements** [comment]
cc @mcdonnelldean

**Comment by mcdonnelldean** [comment]
Landed in #31

**Comment by mcollina** [comment]
I don't think this is a bug or a problem. If you are using a token/password in a pattern, you are in big trouble anyway. I'm 👎 if this causes any perf hit.

**Comment by mcdonnelldean** [comment]
@mcollina Just to note, not only does it not cause a perf hit, it seems to be faster on average, see #31 I added the benchmarks to show. It's not faster by much it is on average faster (on my machine of course)

**Comment by mcollina** [comment]
So, 👍 .

**Comment by mcdonnelldean** [comment]
:D

**Comment by mcdonnelldean** [comment]
Closed via #31
